### PR TITLE
AJ-1971: getWorkspaceId respects Rawls vs. WDS workspace types

### PR DESCRIPTION
**Pull request by davidangb**
## Description

This PR modifies the logic in `CollectionService.getWorkspaceId()` such that virtual collections are only allowed in Rawls-powered workspaces. The logic is:

If workspace is WDS, return the value from sys_wds.collection or throw an error if no row exists. The virtual collections setting is ignored.

If workspace is Rawls, return the value from sys_wds.collection if one exists. Else, return workspaceId == collectionId if virtual collections are allowed. Else, throw an error.

`CollectionService.getWorkspaceId()` continues to respect the `twds.tenancy.enforce-collections-match-workspace-id` config setting. If that setting is true, the method enforces that any workspaceIds returned must match the `$WORKSPACE_ID` env var. This logic hasn't changed; it exists to enforce single-tenancy.

## Questions and Notes for Reviewers

1. `CollectionService.getWorkspaceId()` is a superset of `CollectionService.validateCollection()`. I am not touching `validateCollection()` in this PR, even though there's overlap - in yet one more follow-on PR I will tackle that.
2. I've taken two approaches to fixing unit tests - one "real" approach of inserting into the db and another with mocks. Which do you prefer? See comments inline for details.

### CODE REVIEW

File: service/src/main/java/org/databiosphere/workspacedataservice/service/CollectionService.java [hunk 1]
```diff
@@ -25,38 +26,43 @@
...
-  private final ActivityLogger activityLogger;
-  private final TenancyProperties tenancyProperties;
-  private final TwdsProperties twdsProperties;
 
+  private final ActivityLogger activityLogger;
   private final CollectionRepository collectionRepository;
+  private final DataTableTypeInspector dataTableTypeInspector;
   private final NamedParameterJdbcTemplate namedTemplate;
+  private final TenancyProperties tenancyProperties;
+  private final TwdsProperties twdsProperties;
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/service/CollectionService.java:47]: I alphabetized these here and in the constructor

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java [hunk 2]
```diff
@@ -144,6 +146,11 @@ class PfbQuartzJobControlPlaneE2ETest {
   void setup() {
     collectionId = UUID.randomUUID();
 
+    // do not create a collection; we want to test virtual collections here.
+    // mock out the DataTableTypeInspector to show that this workspace is Rawls-powered.
+    when(dataTableTypeInspector.getWorkspaceDataTableType(WorkspaceId.of(collectionId)))
+        .thenReturn(WorkspaceDataTableType.RAWLS);
+
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java:153]: This approach uses a mock for DataTableTypeInspector to indicate the workspace is Rawls-powered and thus supports virtual collections. Do you like this mocked approach better, as compared to `RawlsJsonQuartzJobControlPlaneE2ETest` just below?

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/rawlsjson/RawlsJsonQuartzJobTest.java [hunk 3]
```diff
@@ -108,7 +118,11 @@ private void assertPubSubMessage(Map<String, String> expectedMessage) {
   private void assertSingleBlobWritten(String expectedBlobName) {
     List<Blob> blobsWritten = listBlobs(storage);
     assertThat(blobsWritten).hasSize(1);
-    assertThat(blobsWritten.get(0).getName()).isEqualTo(expectedBlobName);
+    assertThat(blobsWritten.get(0).getName())
+        .withFailMessage(
+            "Expected blob name is incorrect; expected %s got %s",
+            expectedBlobName, blobsWritten.get(0).getName())
+        .isEqualTo(expectedBlobName);
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/dataimport/rawlsjson/RawlsJsonQuartzJobTest.java:125]: no functional change here; I added a custom failure message to help me when I was debugging failures in this test

File: service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceNoWorkspaceTest.java
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceNoWorkspaceTest.java:1]: this test was obsolete

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/rawlsjson/RawlsJsonQuartzJobControlPlaneE2ETest.java [hunk 4]
```diff
@@ -67,15 +76,23 @@ class RawlsJsonQuartzJobControlPlaneE2ETest {
...
 
   @BeforeEach
   void setup() {
+    TestUtils.cleanAllWorkspaces(namedTemplate);
     collectionId = UUID.randomUUID();
+    workspaceId = WorkspaceId.of(collectionId);
+    // do not create a collection; we want to test virtual collections here.
+    // however, create a record of the corresponding workspace indicating the workspace is
+    // Rawls-powered, so we know this workspace is valid for virtual collections.
+    workspaceRepository.save(new WorkspaceRecord(workspaceId, WorkspaceDataTableType.RAWLS, true));
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/dataimport/rawlsjson/RawlsJsonQuartzJobControlPlaneE2ETest.java:89]: This approach uses no mocks; instead it relies on setting up the appropriate data in the db to indicate the workspace is Rawls-powered and thus supports virtual collections. Do you like this mocked approach better, as compared to `PfbQuartzJobControlPlaneE2ETest` above?

File: service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceGetWorkspaceIdTest.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package org.databiosphere.workspacedataservice.service;
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceGetWorkspaceIdTest.java:1]: I added a few more test cases to this file

File: service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceNoWorkspaceTest.java [hunk 6]
```diff
@@ -1,50 +0,0 @@
-package org.databiosphere.workspacedataservice.service;
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceNoWorkspaceTest.java:1]: this test was obsolete

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java (hunk 2 above)
Comment (calypsomatic) [reply service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java:153]: I think the mocked approach is good, assuming we've tested the code that checks whether the workspaceId is in the DB somewhere else.

**Review by kevinmarete — APPROVED** [review]
Looks good

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java (hunk 2 above)
Comment (davidangb) [reply service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java:153]: cool, thank you - moved tests over to the mock approach in c45fda46a99d0f2c378130ffeb9f2799e440ecf5

**Review by calypsomatic — APPROVED** [review]

**Pull request merged**